File: common/src/main/java/wile/redstonepen/detail/ModRenderers.java
/*
 * @file ModRenderers.java
 * @author Stefan Wilhelm (wile)
 * @copyright (C) 2020 Stefan Wilhelm
 * @license MIT (see https://opensource.org/licenses/MIT)
 */
package wile.redstonepen.detail;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexConsumer;
import net.minecraft.client.renderer.*;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.client.renderer.texture.OverlayTexture;
import net.minecraft.client.Minecraft;
import net.minecraft.client.resources.model.BakedModel;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.Vec3;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import wile.redstonepen.ModConstants;
import wile.redstonepen.blocks.track.RedstoneTrackDefs;
import wile.redstonepen.blocks.track.RedstoneTrackDefs.connections;
import wile.redstonepen.blocks.track.TrackBlockEntity;
import wile.redstonepen.libmc.Auxiliaries;
import wile.redstonepen.libmc.PlatformServices;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.WeakHashMap;


public class ModRenderers
{
  @Environment(EnvType.CLIENT)
  public static class TrackTer implements BlockEntityRenderer<TrackBlockEntity>
  {
    private static final ResourceLocation[] model_rls  = new ResourceLocation[RedstoneTrackDefs.STATE_FLAG_WIR_COUNT];
    private static final ResourceLocation[] modelm_rls = new ResourceLocation[RedstoneTrackDefs.STATE_FLAG_CON_COUNT];
    private static final ResourceLocation[] modelc_rls = new ResourceLocation[RedstoneTrackDefs.STATE_FLAG_CON_COUNT];
    private static final ArrayList<Vec3> power_rgb = new ArrayList<>();
    private final WeakHashMap<TrackBlockEntity, Long> broken_entities_ = new WeakHashMap<>();
    private final BlockEntityRendererProvider.Context renderer_;

    public static List<ResourceLocation> registerModels()
    {
      List<ResourceLocation> resources_to_register = new ArrayList<>();

      RedstoneTrackDefs.models.STATE_WIRE_MAPPING.entrySet().forEach((kv->{
        final ResourceLocation mrl = ResourceLocation.tryBuild(ModConstants.MODID, kv.getValue()).withPrefix("item/");
        for(int i=0; i<RedstoneTrackDefs.STATE_FLAG_WIR_COUNT; ++i) {
          if((kv.getKey() & (1L<<(RedstoneTrackDefs.STATE_FLAG_WIR_POS+i))) != 0) {
            model_rls[i] = mrl;
            break;
          }
        }
        resources_to_register.add(mrl);
      }));
      RedstoneTrackDefs.models.STATE_CONNECT_MAPPING.entrySet().forEach((kv->{
        ResourceLocation mrl = ResourceLocation.tryBuild(ModConstants.MODID, kv.getValue()).withPrefix("item/");
        for(int i=0; i<RedstoneTrackDefs.STATE_FLAG_CON_COUNT; ++i) {
          if((kv.getKey() & (1L<<(RedstoneTrackDefs.STATE_FLAG_CON_POS+i))) != 0) {
            modelc_rls[i] = mrl;
            break;
          }
        }
        resources_to_register.add(mrl);
      }));
      RedstoneTrackDefs.models.STATE_CNTWIRE_MAPPING.entrySet().forEach((kv->{
        ResourceLocation mrl = ResourceLocation.tryBuild(ModConstants.MODID, kv.getValue()).withPrefix("item/");
        for(int i=0; i<RedstoneTrackDefs.STATE_FLAG_CON_COUNT; ++i) {
          if((kv.getKey() & (1L<<(RedstoneTrackDefs.STATE_FLAG_CON_POS+i))) != 0) {
            modelm_rls[i] = mrl;
            break;
          }
        }
        resources_to_register.add(mrl);
      }));
      power_rgb.clear();
      for(int i = 0; i <= 15; ++i) {
        float f = (float)i / 15.0f;
        power_rgb.add(new Vec3(
          Mth.clamp(0.01f + f, 0.0F, 1f),
          Mth.clamp(0.01f + f * 0.4f-.3f, 0.0F, 1f),
          Mth.clamp(0.01f + f * 0.4f-.2f, 0.0F, 1f)
        ));
      }
      return resources_to_register;
    }

    private static Vec3 getPowerRGB(int p)
    { return power_rgb.get(p & 0xf); }

    public TrackTer(BlockEntityRendererProvider.Context renderer)
    { this.renderer_ = renderer; }

    @Override
    @SuppressWarnings("deprecation")
    public void render(final TrackBlockEntity te, float unused1, PoseStack mxs, MultiBufferSource buf, int combinedLightIn, int combinedOverlayIn)
    {
      final long current_flags = te.getStateFlags();
      if(broken_entities_.getOrDefault(te, ~current_flags) == current_flags) return;
      mxs.pushPose();
      try {
        final BlockState block_state = te.getBlockState();
        final VertexConsumer vxb = buf.getBuffer(ItemBlockRenderTypes.getRenderType(block_state, false));
        combinedOverlayIn = OverlayTexture.pack(0, 0);
        {
          final int wirfl = te.getWireFlags();
          final int wirfc = te.getWireFlagCount();
          long flag = 0x1;
          for(int i=0; i<wirfc; ++i, flag<<=1) {
            if((wirfl & flag) == 0) continue;
            final Vec3 rgb = getPowerRGB(te.getSidePower(connections.CONNECTION_BIT_ORDER[i/4]));
            final BakedModel model = PlatformServices.getRendering().getBakedModel(Minecraft.getInstance().getModelManager(), model_rls[i]);
            Minecraft.getInstance().getBlockRenderer().getModelRenderer().renderModel(
              mxs.last(),
              vxb,
              null,
              model,
              (float)rgb.x(), (float)rgb.y(), (float)rgb.z(),
              combinedLightIn,
              combinedOverlayIn
            );
          }
        }
        {
          final int wirfl = te.getWireFlags();
          final int confl = te.getConnectionFlags();
          final int confc = te.getConnectionFlagCount();
          long wir = 0xfL;
          long con = 0x1L;
          for(int i=0; i<confc; ++i, con<<=1, wir<<=4) {
            if(((wirfl & wir)==0) && ((confl & con)==0)) continue;
            final Vec3 rgb = getPowerRGB(te.getSidePower(connections.CONNECTION_BIT_ORDER[i]));
            final BakedModel model = ((confl & con)==0)
              ? PlatformServices.getRendering().getBakedModel(Minecraft.getInstance().getModelManager(), modelm_rls[i])
              : PlatformServices.getRendering().getBakedModel(Minecraft.getInstance().getModelManager(), modelc_rls[i]);
            Minecraft.getInstance().getBlockRenderer().getModelRenderer().renderModel(
              mxs.last(),
              vxb,
              null,
              model,
              (float)rgb.x(), (float)rgb.y(), (float)rgb.z(),
              combinedLightIn,
              combinedOverlayIn
            );
          }
        }
      } catch(Throwable e) {
        if(!broken_entities_.containsKey(te)) {
          Auxiliaries.logError("TER render error for track at " + te.getBlockPos() + ", exception: " + e.getMessage());
          Auxiliaries.logError(String.join("\n", Arrays.stream(e.getStackTrace()).map(StackTraceElement::toString).toList()));
        }
        broken_entities_.put(te, current_flags);
      } finally {
        mxs.popPose();
      }
    }
  }

}

File: common/src/main/kotlin/wile/redstonepen/detail/ModRenderers.kt
package wile.redstonepen.detail

import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexConsumer
import java.util.WeakHashMap
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.ItemBlockRenderTypes
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.resources.ResourceLocation
import net.minecraft.util.Mth
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.Vec3
import wile.redstonepen.ModConstants
import wile.redstonepen.blocks.track.RedstoneTrackDefs
import wile.redstonepen.blocks.track.TrackBlockEntity
import wile.redstonepen.platform.PlatformServices
import wile.redstonepen.util.Auxiliaries

object ModRenderers {

    @Environment(EnvType.CLIENT)
    class TrackTer(ctx: BlockEntityRendererProvider.Context?) :
        BlockEntityRenderer<TrackBlockEntity> {

        companion object {
            private val model_rls =
                arrayOfNulls<ResourceLocation>(RedstoneTrackDefs.STATE_FLAG_WIR_COUNT)
            private val modelm_rls =
                arrayOfNulls<ResourceLocation>(RedstoneTrackDefs.STATE_FLAG_CON_COUNT)
            private val modelc_rls =
                arrayOfNulls<ResourceLocation>(RedstoneTrackDefs.STATE_FLAG_CON_COUNT)
            private val power_rgb = ArrayList<Vec3>()

            @JvmStatic
            fun registerModels(): List<ResourceLocation> {
                val resourcesToRegister = ArrayList<ResourceLocation>()
                RedstoneTrackDefs.Models.STATE_WIRE_MAPPING.forEach { (key, value) ->
                    val mrl =
                        requireNotNull(ResourceLocation.tryBuild(ModConstants.MODID, value))
                            .withPrefix("item/")
                    for (i in 0 until RedstoneTrackDefs.STATE_FLAG_WIR_COUNT) {
                        if ((key and (1L shl (RedstoneTrackDefs.STATE_FLAG_WIR_POS + i))) != 0L) {
                            model_rls[i] = mrl
                            break
                        }
                    }
                    resourcesToRegister.add(mrl)
                }
                RedstoneTrackDefs.Models.STATE_CONNECT_MAPPING.forEach { (key, value) ->
                    val mrl =
                        requireNotNull(ResourceLocation.tryBuild(ModConstants.MODID, value))
                            .withPrefix("item/")
                    for (i in 0 until RedstoneTrackDefs.STATE_FLAG_CON_COUNT) {
                        if ((key and (1L shl (RedstoneTrackDefs.STATE_FLAG_CON_POS + i))) != 0L) {
                            modelc_rls[i] = mrl
                            break
                        }
                    }
                    resourcesToRegister.add(mrl)
                }
                RedstoneTrackDefs.Models.STATE_CNTWIRE_MAPPING.forEach { (key, value) ->
                    val mrl =
                        requireNotNull(ResourceLocation.tryBuild(ModConstants.MODID, value))
                            .withPrefix("item/")
                    for (i in 0 until RedstoneTrackDefs.STATE_FLAG_CON_COUNT) {
                        if ((key and (1L shl (RedstoneTrackDefs.STATE_FLAG_CON_POS + i))) != 0L) {
                            modelm_rls[i] = mrl
                            break
                        }
                    }
                    resourcesToRegister.add(mrl)
                }
                power_rgb.clear()
                for (i in 0..15) {
                    val f = i.toFloat() / 15.0f
                    power_rgb.add(
                        Vec3(
                            Mth.clamp(0.01f + f, 0.0f, 1.0f).toDouble(),
                            Mth.clamp(0.01f + f * 0.4f - 0.3f, 0.0f, 1.0f).toDouble(),
                            Mth.clamp(0.01f + f * 0.4f - 0.2f, 0.0f, 1.0f).toDouble(),
                        )
                    )
                }
                return resourcesToRegister
            }

            private fun getPowerRGB(p: Int): Vec3 = power_rgb[p and 0xf]
        }

        private val brokenEntities = WeakHashMap<TrackBlockEntity, Long>()

        @Suppress("DEPRECATION")
        override fun render(
            te: TrackBlockEntity,
            unused1: Float,
            mxs: PoseStack,
            buf: MultiBufferSource,
            combinedLightIn: Int,
            combinedOverlayIn: Int,
        ) {
            val currentFlags = te.getStateFlags()
            if (brokenEntities.getOrDefault(te, currentFlags.inv()) == currentFlags) return
            mxs.pushPose()
            try {
                val blockState: BlockState = te.blockState
                val vxb: VertexConsumer =
                    buf.getBuffer(ItemBlockRenderTypes.getRenderType(blockState, false))
                val overlay = OverlayTexture.pack(0, 0)
                run {
                    val wirfl = te.getWireFlags()
                    val wirfc = te.getWireFlagCount()
                    var flag = 0x1L
                    for (i in 0 until wirfc) {
                        if ((wirfl.toLong() and flag) != 0L) {
                            val rgb =
                                getPowerRGB(
                                    te.getSidePower(
                                        RedstoneTrackDefs.Connections.CONNECTION_BIT_ORDER[i / 4]
                                    )
                                )
                            val model =
                                PlatformServices.getRendering()
                                    .getBakedModel(
                                        Minecraft.getInstance().modelManager,
                                        requireNotNull(model_rls[i]),
                                    )
                            Minecraft.getInstance()
                                .blockRenderer
                                .modelRenderer
                                .renderModel(
                                    mxs.last(),
                                    vxb,
                                    null,
                                    model,
                                    rgb.x().toFloat(),
                                    rgb.y().toFloat(),
                                    rgb.z().toFloat(),
                                    combinedLightIn,
                                    overlay,
                                )
                        }
                        flag = flag shl 1
                    }
                }
                run {
                    val wirfl = te.getWireFlags()
                    val confl = te.getConnectionFlags()
                    val confc = te.getConnectionFlagCount()
                    var wir = 0xfL
                    var con = 0x1L
                    for (i in 0 until confc) {
                        if ((wirfl.toLong() and wir) != 0L || (confl.toLong() and con) != 0L) {
                            val rgb =
                                getPowerRGB(
                                    te.getSidePower(
                                        RedstoneTrackDefs.Connections.CONNECTION_BIT_ORDER[i]
                                    )
                                )
                            val model =
                                if ((confl.toLong() and con) == 0L) {
                                    PlatformServices.getRendering()
                                        .getBakedModel(
                                            Minecraft.getInstance().modelManager,
                                            requireNotNull(modelm_rls[i]),
                                        )
                                } else {
                                    PlatformServices.getRendering()
                                        .getBakedModel(
                                            Minecraft.getInstance().modelManager,
                                            requireNotNull(modelc_rls[i]),
                                        )
                                }
                            Minecraft.getInstance()
                                .blockRenderer
                                .modelRenderer
                                .renderModel(
                                    mxs.last(),
                                    vxb,
                                    null,
                                    model,
                                    rgb.x().toFloat(),
                                    rgb.y().toFloat(),
                                    rgb.z().toFloat(),
                                    combinedLightIn,
                                    overlay,
                                )
                        }
                        con = con shl 1
                        wir = wir shl 4
                    }
                }
            } catch (e: Throwable) {
                if (!brokenEntities.containsKey(te)) {
                    Auxiliaries.logError(
                        "TER render error for track at ${te.blockPos}, exception: ${e.message}"
                    )
                    Auxiliaries.logError(e.stackTrace.joinToString("\n") { it.toString() })
                }
                brokenEntities[te] = currentFlags
            } finally {
                mxs.popPose()
            }
        }
    }
}
